# Task
Build a classification model for water potability prediction using the `water_potability.csv` dataset.

## Load Data

### Subtask:
Load the `water_potability.csv` dataset into a pandas DataFrame.


**Reasoning**:
Load the `water_potability.csv` dataset into a pandas DataFrame and display the first 5 rows.



In [ ]:
import pandas as pd

df = pd.read_csv('water_potability.csv')
df.head(20)

,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
0,NaN,204.890455,20791.318981,7.300212,368.516441,564.308654,10.379783,86.990970,2.963135,0
1,3.716080,129.422921,18630.057858,6.635246,NaN,592.885359,15.180013,56.329076,4.500656,0
2,8.099124,224.236259,19909.541732,9.275884,NaN,418.606213,16.868637,66.420093,3.055934,0
3,8.316766,214.373394,22018.417441,8.059332,356.886136,363.266516,18.436524,100.341674,4.628771,0
4,9.092223,181.101509,17978.986339,6.546600,310.135738,398.410813,11.558279,31.997993,4.075075,0
5,5.584087,188.313324,28748.687739,7.544869,326.678363,280.467916,8.399735,54.917862,2.559708,0
6,10.223862,248.071735,28749.716544,7.513408,393.663396,283.651634,13.789695,84.603556,2.672989,0
7,8.635849,203.361523,13672.091764,4.563009,303.309771,474.607645,12.363817,62.798309,4.401425,0
8,NaN,118.988579,14285.583854,7.804174,268.646941,389.375566,12.706049,53.928846,3.595017,0
9,11.180284,227.231469,25484.508491,9.077200,404.041635,563.885481,17.927806,71.976601,4.370562,0


## Exploratory Data Analysis (EDA)

### Subtask:
Perform initial data exploration, check for missing values, and understand data distribution.


**Reasoning**:
To perform initial data exploration, I will check for missing values, display descriptive statistics for numerical columns, and examine the distribution of the target variable 'Potability'.



In [ ]:
print("Missing values:\n", df.isnull().sum())
print("\nDescriptive statistics:\n", df.describe())
print("\nDistribution of Potability:\n", df['Potability'].value_counts())

Missing values:
 ph                 491
Hardness             0
Solids               0
Chloramines          0
Sulfate            781
Conductivity         0
Organic_carbon       0
Trihalomethanes    162
Turbidity            0
Potability           0
dtype: int64

Descriptive statistics:
                 ph     Hardness        Solids  Chloramines      Sulfate  \
count  2785.000000  3276.000000   3276.000000  3276.000000  2495.000000   
mean      7.080795   196.369496  22014.092526     7.122277   333.775777   
std       1.594320    32.879761   8768.570828     1.583085    41.416840   
min       0.000000    47.432000    320.942611     0.352000   129.000000   
25%       6.093092   176.850538  15666.690297     6.127421   307.699498   
50%       7.036752   196.967627  20927.833607     7.130299   333.073546   
75%       8.062066   216.667456  27332.762127     8.114887   359.950170   
max      14.000000   323.124000  61227.196008    13.127000   481.030642   

       Conductivity  Organic_carbon  T

## Data Preprocessing

### Subtask:
Handle missing values, encode categorical features (if any), scale numerical features, and split the data into training and testing sets.


**Reasoning**:
The first step in data preprocessing is to handle missing values by filling them with the median for specified columns, as instructed. This will ensure that our dataset is complete before further processing.



In [ ]:
for col in ['ph', 'Sulfate', 'Trihalomethanes']:
    df[col].fillna(df[col].median(), inplace=True)

print("Missing values after imputation:\n", df.isnull().sum())

Missing values after imputation:
 ph                 0
Hardness           0
Solids             0
Chloramines        0
Sulfate            0
Conductivity       0
Organic_carbon     0
Trihalomethanes    0
Turbidity          0
Potability         0
dtype: int64


/tmp/ipython-input-1330259382.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(df[col].median(), inplace=True)


In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Make predictions on the scaled test data
y_pred = model.predict(X_test_scaled)

# Calculate evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Print the evaluation metrics
print(f"Model Accuracy: {accuracy:.4f}")
print(f"Model Precision: {precision:.4f}")
print(f"Model Recall: {recall:.4f}")
print(f"Model F1-Score: {f1:.4f}")

Model Accuracy: 0.6738
Model Precision: 0.5974
Model Recall: 0.3770
Model F1-Score: 0.4623


## Save Model

### Subtask:
Save the trained classification model as a `.pkl` file for deployment.


## Summary:

### Data Analysis Key Findings

*   The `water_potability.csv` dataset was successfully loaded, containing columns such as `ph`, `Hardness`, `Sulfate`, and `Potability`.
*   Initial data exploration revealed significant missing values in key columns: `ph` (491 missing), `Sulfate` (781 missing), and `Trihalomethanes` (162 missing).
*   The target variable, `Potability`, exhibited class imbalance, with 1998 samples classified as not potable (0) and 1278 samples classified as potable (1).
*   During preprocessing, missing values in `ph`, `Sulfate`, and `Trihalomethanes` were successfully imputed using their respective medians, and numerical features were scaled using `StandardScaler`.
*   A `RandomForestClassifier` model was trained and evaluated, achieving the following performance metrics on the test set:
    *   Accuracy: 0.6738
    *   Precision: 0.5974
    *   Recall: 0.3770
    *   F1-Score: 0.4623
*   The model's relatively low recall (0.3770) indicates that it struggled to correctly identify a significant portion of the actual potable water samples.

### Insights or Next Steps

*   The observed class imbalance in the `Potability` target variable likely contributed to the model's lower recall for the positive class. Techniques like oversampling, undersampling, or using Synthetic Minority Over-sampling Technique (SMOTE) could be explored to balance the dataset.
*   Further model optimization, including hyperparameter tuning for the `RandomForestClassifier` or experimenting with alternative classification algorithms (e.g., Gradient Boosting, Support Vector Machines), could improve the model's overall performance, especially recall and F1-score.


## Data Preprocessing

### Subtask:
Handle missing values, encode categorical features (if any), scale numerical features, and split the data into training and testing sets.


## Train Classification Model

### Subtask:
Evaluate the trained classification model using appropriate metrics on the testing data.


**Reasoning**:
The data preprocessing step is complete. The next logical step is to train a machine learning model using the preprocessed training data. I will use a RandomForestClassifier, which is a robust model for classification tasks, and print a confirmation of the model training.


**Reasoning**:
To evaluate the trained model, I will first make predictions on the scaled test data and then calculate and print common classification metrics such as accuracy, precision, recall, and F1-score.


In [ ]:
import pickle

# Save the trained model
model_filename = 'random_forest_model.pkl'
with open(model_filename, 'wb') as file:
    pickle.dump(model, file)

print(f"Trained model saved as {model_filename}")

# Save the scaler object
scaler_filename = 'standard_scaler.pkl'
with open(scaler_filename, 'wb') as file:
    pickle.dump(scaler, file)

print(f"Scaler object saved as {scaler_filename}")

Trained model saved as random_forest_model.pkl
Scaler object saved as standard_scaler.pkl


# Task
The trained `RandomForestClassifier` model and `StandardScaler` object have been saved as `random_forest_model.pkl` and `standard_scaler.pkl` respectively, for deployment.

## Save Model and Scaler

### Subtask:
Save the trained RandomForestClassifier model and the StandardScaler object as .pkl files for deployment.


## Summary:

### Data Analysis Key Findings
*   The trained `RandomForestClassifier` model was successfully saved as `random_forest_model.pkl`.
*   The `StandardScaler` object was successfully saved as `standard_scaler.pkl`.

### Insights or Next Steps
*   The saved model and scaler files are now ready for integration into a deployment environment for making predictions on new data.


## Data Preprocessing

### Subtask:
Handle missing values, encode categorical features (if any), scale numerical features, and split the data into training and testing sets.


## Train Classification Model

### Subtask:
Evaluate the trained classification model using appropriate metrics on the testing data.


**Reasoning**:
The data preprocessing step is complete. The next logical step is to train a machine learning model using the preprocessed training data. I will use a RandomForestClassifier, which is a robust model for classification tasks, and print a confirmation of the model training.


**Reasoning**:
To evaluate the trained model, I will first make predictions on the scaled test data and then calculate and print common classification metrics such as accuracy, precision, recall, and F1-score.


**Reasoning**:
The previous code failed due to a typo in the column name 'Trihalomethames'. I will correct this to 'Trihalomethanes' and then proceed with separating features and target, splitting the data into training and testing sets, and scaling the numerical features as per the instructions.


**Reasoning**:
The data preprocessing step is complete. The next logical step is to train a machine learning model using the preprocessed training data. I will use a RandomForestClassifier, which is a robust model for classification tasks, and print a confirmation of the model training.


## Train Classification Model

### Subtask:
Evaluate the trained classification model using appropriate metrics on the testing data.


**Reasoning**:
To evaluate the trained model, I will first make predictions on the scaled test data and then calculate and print common classification metrics such as accuracy, precision, recall, and F1-score.


## Save Model

### Subtask:
Save the trained classification model as a `.pkl` file for deployment.


## Summary:

### Data Analysis Key Findings

*   The `water_potability.csv` dataset was successfully loaded, containing columns such as `ph`, `Hardness`, `Sulfate`, and `Potability`.
*   Initial data exploration revealed significant missing values in key columns: `ph` (491 missing), `Sulfate` (781 missing), and `Trihalomethanes` (162 missing).
*   The target variable, `Potability`, exhibited class imbalance, with 1998 samples classified as not potable (0) and 1278 samples classified as potable (1).
*   During preprocessing, missing values in `ph`, `Sulfate`, and `Trihalomethanes` were successfully imputed using their respective medians, and numerical features were scaled using `StandardScaler`.
*   A `RandomForestClassifier` model was trained and evaluated, achieving the following performance metrics on the test set:
    *   Accuracy: 0.6738
    *   Precision: 0.5974
    *   Recall: 0.3770
    *   F1-Score: 0.4623
*   The model's relatively low recall (0.3770) indicates that it struggled to correctly identify a significant portion of the actual potable water samples.

### Insights or Next Steps

*   The observed class imbalance in the `Potability` target variable likely contributed to the model's lower recall for the positive class. Techniques like oversampling, undersampling, or using Synthetic Minority Over-sampling Technique (SMOTE) could be explored to balance the dataset.
*   Further model optimization, including hyperparameter tuning for the `RandomForestClassifier` or experimenting with alternative classification algorithms (e.g., Gradient Boosting, Support Vector Machines), could improve the model's overall performance, especially recall and F1-score.


## Save Model

### Subtask:
Save the trained classification model as a `.pkl` file for deployment.


## Summary:

### Data Analysis Key Findings

*   The `water_potability.csv` dataset was successfully loaded, containing columns such as `ph`, `Hardness`, `Sulfate`, and `Potability`.
*   Initial data exploration revealed significant missing values in key columns: `ph` (491 missing), `Sulfate` (781 missing), and `Trihalomethanes` (162 missing).
*   The target variable, `Potability`, exhibited class imbalance, with 1998 samples classified as not potable (0) and 1278 samples classified as potable (1).
*   During preprocessing, missing values in `ph`, `Sulfate`, and `Trihalomethanes` were successfully imputed using their respective medians, and numerical features were scaled using `StandardScaler`.
*   A `RandomForestClassifier` model was trained and evaluated, achieving the following performance metrics on the test set:
    *   Accuracy: 0.6738
    *   Precision: 0.5974
    *   Recall: 0.3770
    *   F1-Score: 0.4623
*   The model's relatively low recall (0.3770) indicates that it struggled to correctly identify a significant portion of the actual potable water samples.

### Insights or Next Steps

*   The observed class imbalance in the `Potability` target variable likely contributed to the model's lower recall for the positive class. Techniques like oversampling, undersampling, or using Synthetic Minority Over-sampling Technique (SMOTE) could be explored to balance the dataset.
*   Further model optimization, including hyperparameter tuning for the `RandomForestClassifier` or experimenting with alternative classification algorithms (e.g., Gradient Boosting, Support Vector Machines), could improve the model's overall performance, especially recall and F1-score.


In [ ]:
import pickle
from google.colab import files

# Re-save the trained model to ensure it's present
model_filename = 'random_forest_model.pkl'
with open(model_filename, 'wb') as file:
    pickle.dump(model, file)
print(f"Trained model re-saved as {model_filename}")

# Re-save the scaler object to ensure it's present
scaler_filename = 'standard_scaler.pkl'
with open(scaler_filename, 'wb') as file:
    pickle.dump(scaler, file)
print(f"Scaler object re-saved as {scaler_filename}")

# Now attempt to download the files
files.download(model_filename)
files.download(scaler_filename)

Trained model re-saved as random_forest_model.pkl
Scaler object re-saved as standard_scaler.pkl


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

**Reasoning**:
The previous code failed due to a typo in the column name 'Trihalomethames'. I will correct this to 'Trihalomethanes' and then proceed with separating features and target, splitting the data into training and testing sets, and scaling the numerical features as per the instructions.



In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# 1. Fill missing values
for col in ['ph', 'Sulfate', 'Trihalomethanes']:
    df[col].fillna(df[col].median(), inplace=True)

print("Missing values after imputation:\n", df.isnull().sum())

# 2. Separate features (X) and target (y)
X = df.drop('Potability', axis=1)
y = df['Potability']

# 3. Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 4. Scale numerical features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("\nData preprocessing complete. Shapes of processed data:")
print(f"X_train_scaled shape: {X_train_scaled.shape}")
print(f"X_test_scaled shape: {X_test_scaled.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

Missing values after imputation:
 ph                 0
Hardness           0
Solids             0
Chloramines        0
Sulfate            0
Conductivity       0
Organic_carbon     0
Trihalomethanes    0
Turbidity          0
Potability         0
dtype: int64

Data preprocessing complete. Shapes of processed data:
X_train_scaled shape: (2620, 9)
X_test_scaled shape: (656, 9)
y_train shape: (2620,)
y_test shape: (656,)


/tmp/ipython-input-3709097116.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(df[col].median(), inplace=True)


**Reasoning**:
The data preprocessing step is complete. The next logical step is to train a machine learning model using the preprocessed training data. I will use a RandomForestClassifier, which is a robust model for classification tasks, and print a confirmation of the model training.



In [ ]:
from sklearn.ensemble import RandomForestClassifier

# Initialize and train the RandomForestClassifier model
model = RandomForestClassifier(random_state=42)
model.fit(X_train_scaled, y_train)

print("RandomForestClassifier model trained successfully.")

RandomForestClassifier model trained successfully.


## Train Classification Model

### Subtask:
Evaluate the trained classification model using appropriate metrics on the testing data.


**Reasoning**:
To evaluate the trained model, I will first make predictions on the scaled test data and then calculate and print common classification metrics such as accuracy, precision, recall, and F1-score.



In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Make predictions on the scaled test data
y_pred = model.predict(X_test_scaled)

# Calculate evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Print the evaluation metrics
print(f"Model Accuracy: {accuracy:.4f}")
print(f"Model Precision: {precision:.4f}")
print(f"Model Recall: {recall:.4f}")
print(f"Model F1-Score: {f1:.4f}")

Model Accuracy: 0.6738
Model Precision: 0.5974
Model Recall: 0.3770
Model F1-Score: 0.4623


## Save Model

### Subtask:
Save the trained classification model as a `.pkl` file for deployment.


## Summary:

### Data Analysis Key Findings

*   The `water_potability.csv` dataset was successfully loaded, containing columns such as `ph`, `Hardness`, `Sulfate`, and `Potability`.
*   Initial data exploration revealed significant missing values in key columns: `ph` (491 missing), `Sulfate` (781 missing), and `Trihalomethanes` (162 missing).
*   The target variable, `Potability`, exhibited class imbalance, with 1998 samples classified as not potable (0) and 1278 samples classified as potable (1).
*   During preprocessing, missing values in `ph`, `Sulfate`, and `Trihalomethanes` were successfully imputed using their respective medians, and numerical features were scaled using `StandardScaler`.
*   A `RandomForestClassifier` model was trained and evaluated, achieving the following performance metrics on the test set:
    *   Accuracy: 0.6738
    *   Precision: 0.5974
    *   Recall: 0.3770
    *   F1-Score: 0.4623
*   The model's relatively low recall (0.3770) indicates that it struggled to correctly identify a significant portion of the actual potable water samples.

### Insights or Next Steps

*   The observed class imbalance in the `Potability` target variable likely contributed to the model's lower recall for the positive class. Techniques like oversampling, undersampling, or using Synthetic Minority Over-sampling Technique (SMOTE) could be explored to balance the dataset.
*   Further model optimization, including hyperparameter tuning for the `RandomForestClassifier` or experimenting with alternative classification algorithms (e.g., Gradient Boosting, Support Vector Machines), could improve the model's overall performance, especially recall and F1-score.


# Task
The trained `RandomForestClassifier` model and `StandardScaler` object have been saved as `random_forest_model.pkl` and `standard_scaler.pkl` respectively, for deployment.

## Save Model and Scaler

### Subtask:
Save the trained RandomForestClassifier model and the StandardScaler object as .pkl files for deployment.


## Summary:

### Data Analysis Key Findings
*   The trained `RandomForestClassifier` model was successfully saved as `random_forest_model.pkl`.
*   The `StandardScaler` object was successfully saved as `standard_scaler.pkl`.

### Insights or Next Steps
*   The saved model and scaler files are now ready for integration into a deployment environment for making predictions on new data.


# Task
The trained `RandomForestClassifier` model and `StandardScaler` object have been saved as `random_forest_model.pkl` and `standard_scaler.pkl` respectively, for deployment.

## Summary:

### Data Analysis Key Findings
*   A `RandomForestClassifier` model was successfully saved as `random_forest_model.pkl`.
*   A `StandardScaler` object was successfully saved as `standard_scaler.pkl`.

### Insights or Next Steps
*   These saved model and scaler files are now prepared for deployment, allowing for their use in a production environment to make predictions on new data.
